drivers/sensor/qdec_mcux/qdec_mcux.c

```diff
@@ -0,0 +1,207 @@
+
+	return 0;
+}
+
+static int qdec_mcux_ch_get(const struct device *dev, enum sensor_channel ch,
+			 struct sensor_value *val)
+{
+	struct qdec_mcux_data *data = dev->data;
+
+	switch (ch) {
+	case SENSOR_CHAN_ROTATION:
+		val->val1 = (int64_t)(data->position * 360) /
+			    data->qdec_config.positionModulusValue;
+		val->val2 = 0;
+		break;
+	case SENSOR_CHAN_RPM:
+		val->val1 = data->revolution;
+		val->val2 = 0;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static const struct sensor_driver_api qdec_mcux_api = {
+	.attr_set = &qdec_mcux_attr_set,
+	.attr_get = &qdec_mcux_attr_get,
+	.sample_fetch = &qdec_mcux_fetch,
+	.channel_get = &qdec_mcux_ch_get,
+};
+
+static void init_inputs(const struct device *dev)
+{
+	int i;
+	const struct qdec_mcux_config *config = dev->config;
+
+#ifdef CONFIG_PINCTRL
+	i = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	assert(i == 0);
+#endif
+
+	/* Quadrature Encoder inputs are only accessible via crossbar */
+	XBARA_Init((XBARA_Type *)config->xbar);
+	for (i = 0; i < config->xbar_maps_len; i += 2) {
+		XBARA_SetSignalsConnection((XBARA_Type *)config->xbar,
+					   config->xbar_maps[i],
+					   config->xbar_maps[i + 1]);
+	}
+}
+
+#ifdef CONFIG_PINCTRL
+#define QDEC_MCUX_PINCTRL_DEFINE(n) PINCTRL_DT_INST_DEFINE(n);
+#define QDEC_MCUX_PINCTRL_INIT(n) .pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),
+#else
+#define QDEC_MCUX_PINCTRL_DEFINE(n)
+#define QDEC_MCUX_PINCTRL_INIT(n)
+#endif
+
+#define QDEC_MCUX_INIT(n)							\
+										\
+	static struct qdec_mcux_data qdec_mcux_##n##_data;			\
+										\
+	BUILD_ASSERT((DT_INST_PROP_LEN(n, xbar_maps) % 2) == 0,			\
+			"xbar_maps length must be an even number");		\
+										\
+	QDEC_MCUX_PINCTRL_DEFINE(n)						\
+										\
+	static const struct qdec_mcux_config qdec_mcux_##n##_config = {		\
+		.base = (ENC_Type *)DT_INST_REG_ADDR(n),			\
+		.xbar = (uint32_t)DT_INST_PROP(n, xbar),					\
+		.xbar_maps_len = DT_INST_PROP_LEN(n, xbar_maps),		\
+		.xbar_maps = DT_INST_PROP(n, xbar_maps),			\
+		QDEC_MCUX_PINCTRL_INIT(n)					\
+	};									\
+										\
+	static int qdec_mcux_##n##_init(const struct device *dev)		\
+	{									\
+		const struct qdec_mcux_config *config = dev->config;		\
+		struct qdec_mcux_data *data = dev->data;			\
+										\
+		LOG_DBG("Initializing %s", dev->name);				\
+										\
+		init_inputs(dev);						\
+										\
+		ENC_GetDefaultConfig(&data->qdec_config);			\
+		data->qdec_config.positionModulusValue =			\
+			DT_INST_PROP(n, counts_per_revolution);			\
+		data->qdec_config.revolutionCountCondition =			\
+			kENC_RevolutionCountOnRollOverModulus;			\
+		data->qdec_config.enableModuloCountMode = true;			\
+										\
+		ENC_Init(config->base, &data->qdec_config);			\
+										\
+		/* Update the position counter with initial value. */		\
+		ENC_DoSoftwareLoadInitialPositionValue(config->base);		\
+										\
+		return 0;							\
+	}									\
```

Comment:
Why not have a common init function shared among all instances?

Reply:
There is a **counts_per_revolution** value for each qdec instance. I want to store the value directly in **qdec_config.revolutionCountCondition** to avoid having a duplicate value. But I cannot set the value before `ENC_GetDefaultConfig` is called.

Reply:
You're duplicating an entire function to avoid duplicating counts_per_revolution (a uint32). You can store counts_per_revolution in the config struct similar to the base address.

Reply:
I added the `count_per_revolution` to the config and tested the size.

Move to config with 1 instance:
```
           FLASH:       29148 B        64 MB      0.04%
             RAM:        5440 B        32 MB      0.02%
```
Duplicate init with 1 instance:
```
           FLASH:       29144 B        64 MB      0.04%
             RAM:        5440 B        32 MB      0.02%
```
Move to config with 2 instances:
```
           FLASH:       29308 B        64 MB      0.04%
             RAM:        5504 B        32 MB      0.02%
```
Duplicate init with 2 instances:
```
           FLASH:       29304 B        64 MB      0.04%
             RAM:        5504 B        32 MB      0.02%
```
It is nearly the same size.

Am I missing something?

Reply:
Build with debugging (CONFIG_DEBUG_OPTIMIZATIONS=y) is more efficient with the config approach:

Move to config with 2 instances and debug:
```
           FLASH:       32320 B        64 MB      0.05%
             RAM:        5504 B        32 MB      0.02%
```
Duplicate init with 2 instances and debug:
```
           FLASH:       32352 B        64 MB      0.05%
             RAM:        5504 B        32 MB      0.02%
```